### PR TITLE
Integrates driver framework

### PR DESCRIPTION
**Pull request by ikskuh**

### CODE REVIEW

File: driver/base/DatagramDevice.zig [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+/// Reads a single `datagram` from the device.
+pub fn read(dd: DatagramDevice, datagram: []u8) ReadError!void {
+    if (dd.vtable.readFn) |readFn| {
+        return readFn(dd.object, datagram);
+    } else {
+        return error.Unsupported;
+    }
+}
+
+object: ?*anyopaque,
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:54]: please put struct members at the top of the file, it's difficult to find them in this sea of text

File: driver/base/DatagramDevice.zig [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    pub fn datagram_device(td: *TestDevice) DatagramDevice {
+        return DatagramDevice{
+            .object = td,
+            .vtable = &vtable,
+        };
+    }
+
+    fn connectFn(ctx: ?*anyopaque) ConnectError!void {
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:120]: 🐍

File: driver/base/DatagramDevice.zig [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        const td: *TestDevice = @ptrCast(@alignCast(ctx.?));
+        if (td.connected)
+            return error.DeviceBusy;
+        td.connected = true;
+    }
+
+    fn disconnectFn(ctx: ?*anyopaque) void {
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:127]: 🐍

File: driver/base/DatagramDevice.zig [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        const td: *TestDevice = @ptrCast(@alignCast(ctx.?));
+        if (!td.connected) {
+            std.log.err("disconnect when test device was not connected!", .{});
+        }
+        td.connected = false;
+    }
+
+    fn writeFn(ctx: ?*anyopaque, datagram: []const u8) WriteError!void {
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:135]: 🐍

File: driver/base/DatagramDevice.zig [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            return error.Unsupported;
+        }
+
+        const dg = td.arena.allocator().dupe(u8, datagram) catch return error.IoError;
+        errdefer td.arena.allocator().free(dg);
+
+        td.packets.append(dg) catch return error.IoError;
+    }
+
+    fn readFn(ctx: ?*anyopaque, datagram: []u8) ReadError!void {
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:152]: 🐍

File: driver/base/DatagramDevice.zig [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    } else {
+        return error.Unsupported;
+    }
+}
+
+object: ?*anyopaque,
+vtable: *const VTable,
+
+pub const VTable = struct {
+    connectFn: ?*const fn (?*anyopaque) ConnectError!void,
```
Comment (mattnite) [inline driver/base/DatagramDevice.zig:58]: 🐍

File: driver/base/DigitalIO.zig [hunk 7]
```diff
@@ -0,0 +1,109 @@
+//!
+//! An abstract digital input/output pin.
+//!
+//! Digital I/Os can be used to drive single-wire data
+//!
+
+const std = @import("std");
+
+const DigitalIO = @This();
```
Comment (mattnite) [inline driver/base/DigitalIO.zig:9]: Shouldn't this be `Digital_IO`?

File: driver/display/colors.zig [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    r: u8,
+    g: u8,
+    b: u8,
+
+    pub fn fromRgb(r: u8, g: u8, b: u8) RGB888 {
+        return RGB888{ .r = r, .g = g, .b = b };
+    }
+};
+
+fn makeDefaultColors(comptime Color: type) type {
```
Comment (mattnite) [inline driver/display/colors.zig:29]: 🐍

File: driver/display/ili9488.zig [hunk 9]
```diff
@@ -0,0 +1 @@
+//! TODO: Implement driver for ILI9488
```
Comment (mattnite) [inline driver/display/ili9488.zig:1]: Please don't create empty TODO files, make a ticket instead.

File: driver/display/ssd1306.zig [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+    &.{ 0x00, 0xDB, 0x40 },
+    &.{ 0x00, 0xDA, 0x12 },
+    &.{ 0x00, 0xD3, 0x00 },
+    &.{ 0x00, 0x40 },
+    &.{ 0x00, 0xA4 },
+    &.{ 0x00, 0xAF },
+};
+
+// Fundamental Commands
+test "setContrast" {
```
Comment (mattnite) [inline driver/display/ssd1306.zig:468]: I think this file would be a great candidate for symbol tests. If not, then please rename the tests to match the snake case function names.

File: driver/framework.zig [hunk 11]
```diff
@@ -0,0 +1,64 @@
+//!
```
Comment (mattnite) [inline driver/framework.zig:1]: empty doc comment

File: driver/input/touch/xpt2046.zig [hunk 12]
```diff
@@ -0,0 +1 @@
+//! TODO: Implement XPT2046 driver
```
Comment (mattnite) [inline driver/input/touch/xpt2046.zig:1]: Please delete empty TODO file, create ticket instead.

File: driver/wireless/sx1278.zig [hunk 13]
```diff
@@ -0,0 +1,2 @@
+//! TODO: Implement driver for SX1278
```
Comment (mattnite) [inline driver/wireless/sx1278.zig:1]: Please delete empty TODO file, create ticket instead.

**Review by mattnite — CHANGES_REQUESTED** [review]
Looks good, just some cleanup needed.

File: driver/display/ili9488.zig (hunk 9 above)
Comment (ikskuh) [reply driver/display/ili9488.zig]: https://github.com/ZigEmbeddedGroup/microzig/issues/249

File: driver/wireless/sx1278.zig (hunk 13 above)
Comment (ikskuh) [reply driver/wireless/sx1278.zig]: https://github.com/ZigEmbeddedGroup/microzig/issues/248

File: driver/input/touch/xpt2046.zig (hunk 12 above)
Comment (ikskuh) [reply driver/input/touch/xpt2046.zig]: https://github.com/ZigEmbeddedGroup/microzig/issues/247

**Comment by ikskuh** [comment]
Did a pretty extensive clean-up, changed the pattern a bit to use comptime options structs, which are more flexible, more uniform and also more convenient in the end.

Also applied more :snake: to the code

**Pull request merged**